### PR TITLE
Adds mob static for Keeper MoMMIs.

### PR DESCRIPTION
**Pull request by ComicIronic**
Keeper MoMMIs now get a blob of mob-shaped static (or something else) over every mob.
This should help with immulsions and reminding them that they shouldn't interfere.
All specific spriting and the static effect thanks to Skowron.
This is what a human looks like to an MoMMI:
Humanist
And a group of mankeys
Mon-keigh

NOW WITH ALTERNATIVE THINGS:
Blank
Letters
You can choose which you see with a verb.
Ps: I also wrote a general utility proc for getting the average color of an icon. This is used in the letters you see there.

### CODE REVIEW

**Comment by Angelite191** [comment]
APPROVE

**Comment by NigglyWiggly** [comment]
Is the static animated? It would be pretty rad if it is

**Comment by ComicIronic** [comment]
Yes, both the mob static and the human static is animated.
Ironic, that.

**Comment by N3X15** [comment]
DF-style ASCII characters via maptext would also be acceptable

**Comment by clusterfack** [comment]
Controversial tag when

**Comment by Boxe** [comment]
>DF-style ASCII characters via maptext would also be acceptable

this this this this

**Comment by Shaska-** [comment]
Will they still be able to see the persons name?
What about emagged MoMMIs? Do they also only see static?

**Comment by d3athrow** [comment]
@Shaska-  Emagged gets their sight back.

**Comment by 9600bauds** [comment]
Any chance of an animated demonstration?

**Comment by SkowronX** [comment]
staticmob

**Comment by AutismSpider** [comment]
fuck this is an eyesore

**Comment by SkowronX** [comment]
Original without static
http://i.imgur.com/im1M8WX.gif

**Comment by frangi** [comment]
I like nexis' suggestion, letters assigned to all mobs and their color changing based on profession would be neat too.

All that static, while COOL, might be headache inducing. Still might be worth trying

**Comment by AutismSpider** [comment]
All of this is terrible to me.

But if it is going be put it in letter assigned to mob+color base on profession would be better.

**Comment by RemieRichards** [comment]
@ComicIronic @SkowronX 

Mind if I port this to /tg/?

**Comment by SkowronX** [comment]
Fine by me.

**Comment by IratePirate** [comment]
I'm alright with this. If you want to interact with the crew play a human engineer, there's no reason to play as one at the minute.

**Comment by Duny-** [comment]
Maybe make it so MoMMIs can choose between seeing static or the sprite without static if they find it too aggressive to the eye? Or ASCII characters depending on the mob type I like that suggestion even if it sounds less robo-like.

**Comment by Arthurdentist** [comment]
Reminds me of Black Mirror, seems like an interesting suggestion though. Quite nice.

**Comment by d3athrow** [comment]
@RemieRichards Wheres the limb rewrite!

**Comment by RemieRichards** [comment]
@d3athrow I'm waiting on you and N3X15 for that...?

**Comment by d3athrow** [comment]
@RemieRichards 
I DONT UNDERSTAAAND

**Comment by RemieRichards** [comment]
@d3athrow I AM EQUALLY CONFUSE.

**Comment by d3athrow** [comment]
Your limb code for tg you said you'd make a PR for vg so it would be gpl and not agpl

**Comment by RemieRichards** [comment]
@d3athrow and Nexis told me he'd speak with you, and get back to me about what specific parts you wanted? I licensed it out to you guys ages ago if you want to cut out the middle man also.
But yeah last I heard I was waiting on you two.

**Comment by d3athrow** [comment]
AAAAAAAH @N3X15

**Comment by ComicIronic** [comment]
You can take the code, though you might want to wait on the Dorf Fort maptext version.

**Comment by RemieRichards** [comment]
@ComicIronic thankyou, However /tg/ seems pretty happy with the present static version!
If the DF Text version comes to fruition I'll query it with the players as to which they prefer or perhaps even an option per drone.

**Comment by clusterfack** [comment]
Really though what is the benefit here? How does mommis being unable to recognize crewmembers in anyway benefit anyone? 

It definitely makes following law 2 impossible as you cant  tell a hardsuited crewmember from one without, it makes the [fun] of seeing people visit your rebuilt derelict pointless because you cant even tell if its the same person ore different people. And it serves no actual BENEFIT which is my main problem with it.

**Comment by ComicIronic** [comment]
@RemieRichards Code's been updated, feel free to look again and take what you like.

**Comment by DeityLink** [comment]
>those dorf icons
FUND IT

**Comment by Probe1** [comment]
mommmis face when

ib1j2so

**Comment by VelardAmakar** [comment]
Hella dope. Merge when?

**Comment by CptWad** [comment]
:+1:

File: code/__HELPERS/icons.dm [hunk 1]
```diff
@@ -48,4 +85,20 @@ proc/adjust_brightness(var/color, var/value)
 	RGB[1] = Clamp(RGB[1]+value,0,255)
 	RGB[2] = Clamp(RGB[2]+value,0,255)
 	RGB[3] = Clamp(RGB[3]+value,0,255)
-	return rgb(RGB[1],RGB[2],RGB[3])
\ No newline at end of file
+	return rgb(RGB[1],RGB[2],RGB[3])
+
+/proc/AverageColor(var/icon/I)
```
Comment (d3athrow) [inline code/__HELPERS/icons.dm]: oh boy now i can make admin spawned items warp in with a nice color

Comment (ComicIronic) [inline code/__HELPERS/icons.dm]: I was really surprised this wasn't in the graphics library to start with.
At least it might be useful in the future.

**Comment by RemieRichards** [comment]
@ComicIronic Thanks, and just as a few tg peeps expressed annoyance with static too!

**Comment by RemieRichards** [comment]
Also, I'm getting issues with swapping vision modes.
I can display each mode perfectly fine, <i>once</i>
after that some mobs would linger on the previous setting, I managed to fix this, except for humans, who will now *sometimes* linger on the previous setting.

I'm sure it's probably in my implementation, but just a heads up.

File: code/modules/mob/living/living.dm [hunk 2]
```diff
@@ -1,3 +1,23 @@
+/mob/living/New()
+	. = ..()
+	generate_static_overlay()
+	for(var/mob/living/silicon/robot/mommi/MoMMI in player_list)
+		if(MoMMI.keeper && !MoMMI.emagged && !MoMMI.syndicate)
+			if(MoMMI.static_choice in static_overlays)
+				MoMMI.static_overlays.Add(static_overlays[MoMMI.static_choice])
+				MoMMI.client.images.Add(static_overlays[MoMMI.static_choice])
+			else
+				MoMMI.static_overlays.Add(static_overlays["static"])
```
Comment (RemieRichards) [inline code/modules/mob/living/living.dm]:
```
/mob/living/silicon/robot/mommi/generate_static_overlay()
	return
```
Byond: 'Wrong value for list'

I solved this by adding && syndicateOverlays.len to the first (keeper, emagged, syndi) check.

Comment (ComicIronic) [inline code/modules/mob/living/living.dm]: True enough.

**Comment by Uso15** [comment]
How would an Emag'd MoMMI work? Wouldn't it need to know who to follow and what orders to recieve? One static blob is the same as any other.

**Comment by ComicIronic** [comment]
Emagging an MoMMI removes the static overlays.
MoMMIs not on the keeper lawset also lose the overlays.

**Comment by d3athrow** [comment]
Oh boy time to wait 6 months and have tg players be dildos in the general saying either
>lol vg doesnt even have static overlays for their drones

or
>lol vg just steals features constantly like that static overlay for the drones

Every. Time.

**Comment by CptWad** [comment]
just pull it now and avoid it all

**Comment by ComicIronic** [comment]
Merge furst.
Besides, I get credit in the PR and that's all I care about.

**Comment by Duny-** [comment]
Merged when

**Comment by frangi** [comment]
>H
>not U
I also think symbols should be thicker

**Comment by SkowronX** [comment]
Merge or close.

**Comment by AutismSpider** [comment]
Close

**Comment by DeityLink** [comment]
Merge

**Comment by CptWad** [comment]
Merge

**Comment by Arthurdentist** [comment]
merge it

**Comment by Probe1** [comment]
merge it

**Comment by clusterfack** [comment]
I'd like to see all the above people who say 'merge it' forced to play mommi for a week with the static and see how it affects their enjoyment of the game.

**Comment by SkowronX** [comment]
Merge for now.

**Comment by CptWad** [comment]
@clusterfack Sure.

**Comment by Aurx** [comment]
Merge

**Comment by Probe1** [comment]
I put stock in clusterfacks opinion over mine and after consideration will change what I wrote.

**Comment by Shaska-** [comment]
As cool as this is, isn't this just gonna make it harder for MoMMIs to properly follow their laws?
I don't really see any good coming out of that, neither for the crew nor the MoMMIs.

**Comment by Aurx** [comment]
>As cool as this is, isn't this just gonna make it harder for MoMMIs to properly follow their laws?

How so?

**Comment by Shaska-** [comment]
You can tell a lot about a persons intentions by what he's wearing and holding.
For example, if you enter a room through the door and a person is inside the room, you can't tell if he's just someone with access to the room, working there, or if he's someone who has been trapped in there and just set free by your actions. Or perhaps you've unwittingly let in that murderer everyone has been screaming about on the radio.
It's mostly situations where you would previously have been able to make an educated guess about someones intentions, and know how best to stay out of their way, without helping or hindering them.
There is of course also Clusters example. You can't know if opening that firelock, or whatever, is going to harm the person on the other side of it or not, if you can't see if he's wearing a hardsuit.

**Comment by Aurx** [comment]
>For example, if you enter a room through the door and a person is inside the room, you can't tell if he's just someone with access to the room, working there, or if he's someone who has been trapped in there and just set free by your actions. Or perhaps you've unwittingly let in that murderer everyone has been screaming about on the radio.

People tailgate MoMMIs all the time to get into rooms already.  It's not a huge issue for the MoMMI.  

>There is of course also Clusters example. You can't know if opening that firelock, or whatever, is going to harm the person on the other side of it or not, if you can't see if he's wearing a hardsuit.

Can't know if they have their magboots on either.  I've personally opened firelocks with hardsuited people on the other side, only for them to die horribly to ZAS because they didn't have their magboots turned on.  Or sometimes didn't have their internals turned on.  Oh well.  And what if somebody's just offscreen without a hardsuit on?  Or on the other side of an opaque door with no firelock under it?  That's just as much of a lack of information, and we clearly don't punish for that.  
And anyways, if somebody's loitering outside a firelock'd area, common sense would indicate that they're aware of the risks.

**Comment by Probe1** [comment]
I agree with what @Shaska- wrote, especially the first paragraph where you disagreed.  Every time I play MoMMI I make a judgement call like that where I decide if using a door would let someone in and indirectly involve myself in the affairs of the crew.  Is it a scientist standing outside research or an assistant?  Should I ventcrawl to my inconvenience or should I just say fuck it and go in.

Taking away information is taking away information.

**Comment by ComicIronic** [comment]
But deliberately not opening doors isn't something an MoMMI would do, particularly since
1. They should be ventcrawling anyways to get places faster
2. They don't care about the crew
Gaming, particularly metagaming, over who gets in where based on your actions is not your responsibility.

**Comment by Duny-** [comment]
A silicon unaware its laws are being broken isn't breaking its laws.
Just sayan. Such a change actually makes following the horrible laws easier imo.

**Comment by Shaska-** [comment]
>People tailgate MoMMIs all the time to get into rooms already. It's not a huge issue for the MoMMI. 

Most of the tailgating is a non-issue, granted, but I'd argue it starts becoming an issue once MoMMIs unwittingly lets a pair of trapped nuke-ops out of the brig or something to that effect.

And while admins might not punish players for things they couldn't possibly have known, other players most certainly will. A lot of people jump at the chance to smash silicons to bits for anything they perceive as lawbreaking, and if they did technically break their laws, even if they didn't know, there probably won't be much the admins can do about it, unless we also either adopt a policy of what constitutes "acceptable lawbreaking" for MoMMIs or put a flat-out ban on killing blue-eyed MoMMIs as a non-antag.

Honestly, I can see a lot of potential problems all this might cause, but it's hard how many of them, if any, are going to become an actual problem, without a thorough playtest of it. 
I'd suggest trying it for a week or two, then opening a revert request to discuss whether or not it's something we want to keep.


@ComicIronic 
Deliberately not opening a door is something a MoMMI would do, if opening it means they'd become involved in the affairs of another being.
Ventcrawling is not faster if you need to get to the other side of a door.
They care about the affairs of the crew, or more specifically how to avoid becoming involved in those affairs, so to some degree, they do care about the crew.
Basing a law interpretation on what you can see and hear does not constitute metagaming.

**Comment by ComicIronic** [comment]
>They care about the affairs of the crew, or more specifically how to avoid becoming involved in those affairs, so to some degree, they do care about the crew.

But an MoMMI shouldn't care about any of those things. The indirect impacts of their actions are largely meaningless, since even changing their own choices based on the living beings around them constitutes a type of involvement by altering the future. If opening a door or not opening a door can be so important, an MoMMI should logically do what it was always going to do, instead of changing its actions and risking an alternative result based on the living beings it observes.

**Comment by Shaska-** [comment]
I wouldn't say choosing inaction constitutes a form of involvement.
As for the future stuff, a MoMMIs actions should not alter the course of events. 
They should seek to have events continue as if they were never even there, as altering it would essentially mean you involved yourself somewhere down the line.

**Comment by Aurx** [comment]
>As for the future stuff, a MoMMIs actions should not alter the course of events.
They should seek to have events continue as if they were never even there, as altering it would essentially mean you involved yourself somewhere down the line.

Absolutely not the case.  
A MoMMI is expected to power the station, repair damage, construct/refurbish areas, etc.  
All of which alter the course of events.  Pretty dramatically in the case of repairs to major systems/areas or powering the station.

**Comment by Shaska-** [comment]
I was referring to isolated local events mostly, such as in the case of opening doors in the presence of certain people.
MoMMIs are generally allowed to ignore the 1st law when it comes to following the 3rd law, but it does become more prominent when the 3rd law is not involved in a situation

**Comment by SkowronX** [comment]
And now we have confrictu.

**Comment by clusterfack** [comment]
https://github.com/tgstation/-tg-station/issues/8191

**Comment by RemieRichards** [comment]
@clusterfack I don't exactly agree with that issue report, I've never been able to duplicate the error and noone else has reported the issue.

**Comment by dylanstrategie** [comment]
It's really not needed. For one, MoMMIs often act as passive spectators, which SHOULD be fine

And then, if a MoMMI wants to interfere, it doesn't matter if they see a completely outfitted human being or a pulsing dark mob

If anything this would prevent metagruding, but that alone would be the ground for a lengthy silicon and general ban, so that's more of an admin issue

**Comment by ComicIronic** [comment]
>MoMMIs often act as passive spectators, which SHOULD be fine

That's what observing is for.

>And then, if a MoMMI wants to interfere, it doesn't matter if they see a completely outfitted human being or a pulsing dark mob

That's not the point. It won't stop an MoMMI bent on interfering, but it will stop MoMMIs that interfere subtly by changing the way they behave around people.

**Comment by DeityLink** [comment]
MoMMis interfering subtly break my emulsions.

**Comment by AutismSpider** [comment]
Admins job to ban them like dylanstrategie noted.

God fucking help them do this.

**Comment by ComicIronic** [comment]
No conflicto pls.

**Comment by Hoshi-chan** [comment]
One of the big points of ambiguity in Asimov's work (on which the whole 'law' system was based) was whether or not a silicon was aware of the consequences of its actions; so a robot could harm someone, if it wasn't aware that what it was doing was harmful. Sadly, ambiguity is rarely an element of playing as a cyborg or the AI, because you're in constant and direct interaction with the crew, you have a slew of information and feedback available to you, and you cannot take a passive role in upholding your first law. The only way that a silicon could be made to harm humans in this way would be to divide some series of harmless tasks between individual cyborgs, that, together, would create a Rube Goldberg machine of harm, but even that is unlikely since the cyborgs would be in communication over binary and likely puzzle out the intent of the separate requests.

With MoMMIs, though, it's different. There's a totality to their first law: you aren't to involve yourself in the affairs of other beings at all. This means that you don't open doors when people are around, you ventcrawl. If you see a being, you come back later, when it's not there. "Intent" or any other consideration of those beings should NEVER interfere with your logic as a MoMMI. Beings are to be avoided, and if you're not aware of them, or that your actions might be somehow affecting them, then as far as your laws are concerned, you're in the clear. That's why it's fine to start up the supermatter engine, and not give two shits when someone walks in without radiation protection and dies. You weren't aware, and they chose to walk into a dangerous area, so that's on them.

Making "beings" highlighted in this way seems fine if your intent is to demonstrate what you're not allowed to interact with, so long as there are alternatives that aren't eye-searing (which seems to be the case). But if you're going to blind them to distinguishing between beings like this, why not deafen them to beings, as well? Why go through the trouble of making them see people differently when they can still hear them just fine? What's the point?

**Comment by Probe1** [comment]
there exists a mechanism to stop a mommi from doing something by telling it so.  i.e. "Mobile MMI 476 I am going to set up the supermatter.  Do not involve yourself in my work" and it has to scurry off buzzing all the way.

if it was also deaf the only recourse would be to beat it until it left.

**Comment by ComicIronic** [comment]
Ultimately, the reason they aren't deaf is the same reason they aren't made completely blind to mobs: it would probably increase lawbreaking by sheer accident than reduce it in any way.

**Comment by d3athrow** [comment]
Gonna try it for medium RP

**Comment by AutismSpider** [comment]
>this shit gets merged

so people who want it merged going play mommi right or are they going bitch out of it?

my money is on them not playing mommi for the two weeks at all

**Comment by d3athrow** [comment]
Who knows

**Comment by Probe1** [comment]
gay

**Comment by d3athrow** [comment]
>gay

No You

**Comment by Probe1** [comment]
well we'll see in two weeks.

**Comment by d3athrow** [comment]
Yup

**Comment by DeityLink** [comment]
I actually might start playing MoMMi

File: code/__HELPERS/icons.dm (hunk 1 above)
Comment (9600bauds) [inline code/__HELPERS/icons.dm]: It doesn't seem to work properly. I've tested it extensively and it seems to return the blend of the bottom-rightest 2-3 pixels in the sprite, which is almost always a shade of brown due to the item having an outline. I'm working on a fixed version of this, pester me about this on the IRC

**Pull request merged**